File: javascript/0275_H-Index_II.js
/**
 * @param {number[]} citations
 * @return {number}
 */
var hIndex = function(citations) {
    const n = citations.length;
    let left = 1;
    let right = n + 1;

    while (left < right) {
        let mid = (left + right) >> 1;

        if (citations[n - mid] >= mid) {
            left += 1;
        } else {
            right = mid;
        }
        
    }

    return left - 1;
};

File: python/0275_H-Index_II.py
class Solution:
    def hIndex(self, citations: List[int]) -> int:
        left = 1
        right = len(citations) + 1
        
        while left < right:
            mid = (left + right) // 2

            if citations[-mid] >= mid:
                left = mid + 1
            else:
                right = mid
            
        return left - 1